Reduce the number of calls to LookupSpecialMember.

mordante opened the pull request:
Reducing the number of calls has two benefits:
- It does not add unneeded information to the AST.
- Less information to the AST means less AST based unit tests fail.

With these changes only the following unit tests fail:
- Clang :: AST/ast-dump-decl.cpp
- Clang :: AST/ast-dump-record-definition-data-json.cpp
- Clang :: CXX/special/class.copy/p11.0x.move.cpp
- Clang :: CXX/special/class.dtor/p5-implicit.cpp

Quuxplusone committed:
Implement `[[trivially_relocatable]]` and `__is_trivially_relocatable(T)`.

Also implement `__has_extension(trivially_relocatable)` so that the
library can detect whether the feature exists.

Add attribute [[clang::maybe_trivially_relocatable]].
Based on a suggestion by rjmccall.
I don't think this is a *good* idea, but let's see what happens.

Add attribute [[trivially_relocatable(bool)]].

Eliminate the diagnostic for [[trivially_relocatable]] on second declaration.
It should really go on the (single) class *definition*, by analogy with
"trivially copyable." How do we make that happen?

Quuxplusone committed:
Temporarily eliminate the segfault in LookupSpecialMember.

https://bugs.llvm.org/show_bug.cgi?id=38671

Quuxplusone committed:
Rename [[trivially_relocatable]] to [[clang::trivially_relocatable]].

For the Clang review D50119.

Quuxplusone committed:
Revert "Rename [[trivially_relocatable]] to [[clang::trivially_relocatable]]."

For the Godbolt image.

Quuxplusone committed:
Try to diagnose [[trivially_relocatable]] on forward declarations.

Quuxplusone committed:
update unit tests and CE hash

Quuxplusone committed:
Revert "Try to diagnose [[trivially_relocatable]] on forward declarations."

This reverts commit c8f2501b7a870bfd46f44fbba832be957aadb61d.

It doesn't work because `isThisDeclarationADefinition` doesn't exist for
classes and class templates. :(

    lib/Sema/SemaDeclAttr.cpp:6087:11: error: no member named 'isThisDeclarationADefinition' in 'clang::Decl'; did
          you mean 'isThisDeclarationReferenced'?
      if (!D->isThisDeclarationADefinition()) {
              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
              isThisDeclarationReferenced

Quuxplusone committed:
Fix std::move-related build failures with Xcode.

mordante committed:
Reduce the number of calls to LookupSpecialMember.

Reducing the number of calls has two benefits:
- It does not add unneeded information to the AST.
- Less information to the AST means less AST based unit tests fail.

With these changes only the following unit tests fail:
- Clang :: AST/ast-dump-decl.cpp
- Clang :: AST/ast-dump-record-definition-data-json.cpp
- Clang :: CXX/special/class.copy/p11.0x.move.cpp
- Clang :: CXX/special/class.dtor/p5-implicit.cpp

mordante commented:
Arthur this is a smaller patch than the one I mailed you before. I found some ways to reduce the number of additional modifications to clang. 
When I combine pull request #1 and #2 I only have two unit tests failing:
- Clang :: CXX/special/class.copy/p11.0x.move.cpp
- Clang :: CXX/special/class.dtor/p5-implicit.cpp

I haven't looked at them yet, will do so in the future.

Quuxplusone commented on lib/Sema/SemaDeclCXX.cpp:
```diff
@@ -6158,6 +6158,35 @@ static bool canPassInRegisters(Sema &S, CXXRecordDecl *D,
   return HasNonDeletedCopyOrMove;
 }
 
+static bool CheckTrivalDestructor(Sema &S, CXXRecordDecl *R,
```
s/Trival/Trivial/
A bool-returning function named `Check...` strikes me as confusing. I think the meaning here is more like `bool HasTrivialDestructor`, except that `Has...` implies constness, whereas this function reserves the right to modify `*R` in the process of determining the triviality of its destructor.

Since `LookupSpecialMember` has the same semantics — "basically a getter, but might modify its argument anyway" — perhaps this function should be named similarly, something like `LookupTrivialityOfDestructor`.

mordante committed:
Update the validation functions.

Some changes based on Arthur's code review.

- Trival -> Trivial
- CheckFoo -> lookupFoo
  The lowercase is LLVM's coding style. I like lookupTrivialDestructor
  over lookupTrivialityOfDestructor since the former allows to grep for
  TrivialDestructor.
- Invert the logic of the functions; a return of true means the lookup
  succeeded.

Quuxplusone commented:
Unfortunately this patch regresses the following test case. I found this regression only very luckily because of my existing libc++ tests — `is_trivially_relocatable_v<unordered_map<int, int>>` suddenly became `false` after your patch. This is what I got by running `creduce` on that test and hand-tweaking it a bit.

I added this new regression test to this repo in 74dc201b36059137f26e462222130427004e87f7.

```
// T25 is analogous to unique_ptr; T25c is analogous to __compressed_pair.
template<class T>
struct [[clang::maybe_trivially_relocatable]] T25 {
    T25(T25&&);
};
template<class T> struct T25b { T m_a; };
template<class T> struct T25c : T25b<T> {};
static_assert(__is_trivially_relocatable(T25<int>), "it has the attribute");
static_assert(__is_trivially_relocatable(T25b<int>), "all its members are trivially relocatable");
static_assert(__is_trivially_relocatable(T25c<int>), "all its bases are trivially relocatable");
static_assert(__is_trivially_relocatable(T25b<T25<int>>), "all its members are trivially relocatable");
static_assert(__is_trivially_relocatable(T25c<T25<int>>), "all its bases are trivially relocatable");
```

The current version of this PR causes that very last `static_assert` to fail unexpectedly.

mordante commented:
I can confirm the issue and will look into it.
I get the feeling something odd is going on: when I modify your test for `struct T17` to use `T9` instead of `T10` I get the same issue.

mordante committed:
Fixes the regression in the tests.

This fixes regression as shown in
74dc201b36059137f26e462222130427004e87f7

mordante committed:
Merge upstream.

mordante committed:
Merge branch 'trivially-relocatable' into clang

mordante commented:
I found the solution seems I forgot to copy the required modifications of T10 to the MoveConstructor.
Not sure how easy it will be to merge this branch due to the force push.
If it's not easy just let me know and I'll create a new pull request.

Quuxplusone commented:
> Not sure how easy it will be to merge this branch due to the force push.

The easiest thing would be for you to rebase (or squash-merge) your branch on your side, and then force-push your branch. Something like this [UNTESTED]:

    git fetch Quuxplusone  # assuming you have a remote by this name
    git checkout Quuxplusone/trivially-relocatable
    git merge --squash clang
    git checkout -B clang2  # stick a branch label here (could use 'clang' instead of 'clang2' if you don't mind blowing away that branch label's old location)
    git push -f origin clang2:clang  # force-push the result up to your 'clang' branch on GitHub

But I'm fully capable of doing the equivalent dances on my side; and anyway I would want to mess around with your commit message so I wouldn't really be merging directly anyway. You'd do the dance above just to disentangle what's currently shown in the "Commits" tab on GitHub.

mordante commented:
I'll redo everything and make a new pull request. I get conflicts in the unit tests it's easier to redo them than to try to resolve them in the branch. (Since I test in the mono repro I don't feel confident to resolve the conflicts without testing them.)

mordante commented:
Replaced by #3.
